server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,181 @@
+    }
+
+    public static List<NamedXContentRegistry.Entry> getNamedXContentParsers() {
+        return List.of(
+            new NamedXContentRegistry.Entry(Metadata.Custom.class, new ParseField(HealthMetadata.TYPE), HealthMetadata::fromXContent)
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    private HealthMetadata createHealthMetadata() {
+        return new HealthMetadata(
+            HealthMetadata.DiskThresholds.createDiskThresholds(allocationDiskThresholdSettings, healthDiskThresholdSettings)
+        );
+    }
+
+    private void submitHealthMetadata(String source) {
+        submitHealthMetadata(source, 0);
+    }
+
+    // Visible for testing
+    void submitHealthMetadata(String source, int attempt) {
+        if (attempt < MAX_RETRIES) {
+            HealthMetadata localHealthMedata = createHealthMetadata();
+            String message = attempt == 0 ? source : "retry[" + attempt + "]-" + source;
+            var task = new UpdateHealthMetadataTask(localHealthMedata, () -> submitHealthMetadata(source, attempt + 1));
+            var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+            clusterService.submitStateUpdateTask(message, task, config, taskExecutor);
+        } else {
+            logger.error("Failed to process {} after {} attempts", source, attempt);
+        }
+    }
+
+    record UpdateHealthMetadataTask(HealthMetadata healthMetadata, Runnable retry) implements ClusterStateTaskListener {
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "never called";
+        }
+
+        @Override
+        public void onFailure(@Nullable Exception e) {
+            logger.error("unexpected failure during health metadata update", e);
+            retry.run();
+        }
+
+        static class Executor implements ClusterStateTaskExecutor<UpdateHealthMetadataTask> {
+
+            @Override
+            public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateHealthMetadataTask>> taskContexts)
+                throws Exception {
+                final HealthMetadata originalHealthMetadata = HealthMetadata.getHealthCustomMetadata(currentState);
+                HealthMetadata mostRecentHealthMetadata = originalHealthMetadata;
+                if (taskContexts.isEmpty() == false) {
+                    mostRecentHealthMetadata = taskContexts.get(taskContexts.size() - 1).getTask().healthMetadata();
+                    if (Objects.equals(originalHealthMetadata, mostRecentHealthMetadata)) {
+                        mostRecentHealthMetadata = originalHealthMetadata;
+                    }
```

Comment:
This is a bit confusing and if I understand correctly it's a micro optimisation? Do we aim to skip some task executions? 

Would it be simpler to iterate over all tasks and execute them like we do everywhere else? ie. https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L115

Reply:
Kind of yes, since we just pick the latest version and we update the cluster state I thought it makes sense to just pick the last one. On the other hand since we are already iterating the task contexts to signal the success then it is not the optimization I thought it was. I will change it.